### ISSUE TITLE
Facetting a layered plot throws an error

### ISSUE THREAD

**Issue by kuchenrolle** [altair-viz/altair#668]
When I try to facet a layered plot in jupyterlab, altair complains that it's invalidly specified and "'data' is a required property":

```python
import altair as alt
cars = alt.load_dataset('cars')

first = alt.Chart(cars).mark_point().encode(x = "Horsepower", y = "Acceleration")
second = alt.Chart(cars).mark_bar().encode(x = "Horsepower", y = "Acceleration")

combined = first + second # works
first.facet("Cylinders:N") # works
second.facet("Cylinders:N") # works
combined.facet("Cylinders:N") # fails
```
altair version: 2.0.0rc1
jupyterlab version: 0.31.12

**Comment by jakevdp** [comment]
Thanks for the report. Just to help other people find this, the error that comes up is this:
```pytb
---------------------------------------------------------------------------
SchemaValidationError                     Traceback (most recent call last)
~/github/altair-viz/altair/altair/vegalite/v2/api.py in to_dict(self, *args, **kwargs)
    336         if dct is None:
    337             kwargs['validate'] = 'deep'
--> 338             dct = super(TopLevelMixin, copy).to_dict(*args, **kwargs)
    339 
    340         if is_top_level:

~/github/altair-viz/altair/altair/utils/schemapi.py in to_dict(self, validate, ignore, context)
    233                 self.validate(result)
    234             except jsonschema.ValidationError as err:
--> 235                 raise SchemaValidationError(self, err)
    236         return result
    237 

SchemaValidationError: Invalid specification

        altair.vegalite.v2.api.FacetChart, validating 'required'

        'data' is a required property
```
The reason this errors is that the ``FacetSpec`` object created by ``alt.facet`` does not have any data associated with it. You can fix this by providing the data directly:
```
combined.facet("Cylinders:N", data=cars)
```

When working with compound charts like layers or facets, you need to think a bit about where the data is specified; the altair docs mention this briefly, though the discussion could definitely be made more complete: https://altair-viz.github.io/user_guide/compound_charts.html#advanced-compound-charts-and-data-specification

The reason the facets work in the first two cases is that the charts being faceted specify their data, and so the facet function can grab that data and use it.

In the case of the layered chart, the data is not specified at the top level, and so the facet chart does not have any data to reference. If the layer chart does have data defined at the top level, then the facet can access it; you can see that in the fact that this works:
```
combined = alt.layer(first, second, data=cars)
combined.facet("Cylinders:N")
```
Now, we could design ``facet()`` to go digging through the layers to find potential data sources, but that could lead to problems (i.e. what if the two layers have different data sources? Which should take priority?) In the end, we decided that explicit is better than implicit in this case.

There is an open issue thinking about the possibility of having compound charts detect when subcharts have identical data and moving that up to the top level (#579), but I still feel a bit queasy about the assumptions that would require.

Hope that answers your question!

**Comment by kuchenrolle** [comment]
When I supply the data like that, the facetted plots are identical and display all the data. If I try to stick to the documentation, but use the "+" operator, I get the same error as before:

```
base = alt.Chart(cars).encode(x = "Horsepower", y = "Acceleration")
first = base.mark_point()
second = base.mark_bar()
# third = first + second
third = layer(first, second, data=cars)
third.facet("Cylinders:N") # fails with third = first + second; all plots identical if using layer()
third.facet("Cylinders:N", data = cars) # all plots identical either way
```

**Comment by jakevdp** [comment]
Ah, interesting. I hadn't appreciated this behavior before.

So in this example, you're still specifying the data in three places: one in each layer, and one at the top level. Because each layer has its own data, that data overrides the faceted data coming down from the parent chart. What you need to do is create layers that **do not** have their own data, so that they see the data coming from the parent object.

In other words, change ``base = alt.Chart(cars)`` to ``base = alt.Chart()`` (no data specified) and then you'll get different sets of points in each facet.

**Comment by kuchenrolle** [comment]
That works! (Is not all that intuitive, though. But then again, this project is quite new and is really awesome generally.)

Thanks!

**Comment by jakevdp** [comment]
Which part is not intuitive? I'm just curious how we could document the data model for compound charts so that it would be more intuitive.

**Comment by kuchenrolle** [comment]
Well, the part where the same data specified in multiple places results in a different plot from that same data specified in one place. Or that specifying the data in the last place does not simply make data specified in earlier layers irrelevant.

**Comment by jakevdp** [comment]
OK, I see. I think the model makes sense if presented the right way... i.e. parents pass their data down to their children, but the children can override it if they choose.

In the case of the facet chart, the parent facet object passes portions of the data down to the children. But if you specify the full dataset in one of the children, it will use that instead.

It may be that the presence of the ``facet()`` method itself is confusing, because it obscures the fact that a hierarchical chart is being created, and we should instead have people explicitly build a ``FacetChart()`` object.

**Comment by jakevdp** [comment]
This is now addressed in the documentation: https://altair-viz.github.io/user_guide/compound_charts.html#faceted-charts

**Comment by kfrncs** [comment]
I'm having the same problem with data duplication even after removing the data specification in the child alt.Chart()

**Comment by jakevdp** [comment]
You specified the data in two places; the full copy of the data specified in the layer plot overrides the partial copies of data coming from the facet plot.

**Comment by jakevdp** [comment]
More concretely, do this and it should work:
```python
combined = alt.layer(points, fit_line)
combined.facet("baths:N", data=sac_csv)
```

**Comment by kfrncs** [comment]
Thank you so much! That solved it
